### PR TITLE
update public interface to logging

### PR DESCRIPTION
**Pull request by zancas**
Start hacking on serializing to a file

### CODE REVIEW

**Comment by dannasessha** [comment]
Just as a point of order, this force push threw me off just a bit because I wasn't expecting it: I had put some time into reviewing the prior version to then find that it was outdated.

**Comment by dannasessha** [comment]
I'm not sure if it's possible  to see I have started a review before I finish comments, but if not, it would be helpful in the future to have some chatter about force-pushes or continuing development of the branch after it's mentioned in our communication or a review is requested here on github.

**Comment by dannasessha** [comment]
The message I get looking for my comments is now:

"We went looking everywhere, but couldn’t find those commits.

Sometimes commits can disappear after a force-push. Head back to the latest changes here."

**Review by dannasessha — COMMENTED** [review]
Abandoning this review now. Starting a new one momentarily.

**Comment by zancas** [comment]
Ahhh...   hmmm...   yeah, so this is feature branch atop a feature branch with 0 collaborators within the branch.   I hadn't considered the effect on reviews.

Maybe we need a protocol?

I general I've followed the policy of not force-pushing, once code is shared.

File: src/lib.rs [hunk 1]
```diff
@@ -739,6 +767,32 @@ mod unit {
     fn serde_json_value_help_getinfo() {
         let getinfo_serde_json_value = test::getinfo_export();
         let help_getinfo = interpret_help_message(test::HELP_GETINFO);
-        assert_eq!(getinfo_serde_json_value, help_getinfo);
+        assert_eq!(getinfo_serde_json_value, help_getinfo.1);
+    }
+
+    #[test]
```
Comment (dannasessha) [inline src/lib.rs:773]: Okay, nice! I notice all tests are passing, including this one.

Good work!

Comment (dannasessha) [reply src/lib.rs:773]: As a note, I did not have to do `PATH=$PATH:${PATH_TO_ZCASHCLI_DIR} cargo test` to get record_interpretation_getblockchaininfo() to pass, as may have been suggested.

File: src/lib.rs [hunk 2]
```diff
@@ -739,6 +767,32 @@ mod unit {
...
+        );
+        let output = std::path::Path::new(&location);
+        record_interpretation(
+            test_cmd.to_string(),
+            getblockchainfo_interpretation(),
+        );
+
+        //Now let's examine the results!
+        let reader =
+            std::io::BufReader::new(std::fs::File::open(output).unwrap());
```
Comment (dannasessha) [inline src/lib.rs:791]: I am not (yet) familiar with these techniques but I am reading up on them...

**Review by dannasessha — COMMENTED** [review]
Currently I am able to run tests which all pass for me locally.

However, the code panics with `cargo run --bin quizface`.

I believe this is because some of the logic used to limit which commands were being run (`main.rs` line 15 in the older commits) has been removed.

I hope this comments are helpful to advancing discussion about the changes, I have more to learn, especially about the buffering and stream aspects of the new test.

I would like to see this compile, and talk over some of these comments, but overall this looks to be a lot of good work!

File: src/lib.rs [hunk 3]
```diff
@@ -55,10 +55,30 @@ pub fn check_success(output: &std::process::ExitStatus) {
...
     let (cmd_name, result_data) = extract_name_and_result(raw_command_help);
     let scrubbed_result = scrub_result(cmd_name.clone(), result_data);
-    annotate_result(&mut scrubbed_result.chars())
+    (cmd_name, annotate_result(&mut scrubbed_result.chars()))
+}
+
+fn record_interpretation(cmd: String, interpretation: serde_json::Value) {
+    let location = format!(
+        "./output/{}/{}.json",
+        utils::logging::create_version_name(),
```
[the remark this answers is not in the document]
Comment (zancas) [reply src/lib.rs:69]: Well...   in a way a (idempotent) `fn` is like a variable name.  As long as there are no side-effects, it is a Single Source of Truth.

File: src/lib.rs [hunk 4]
```diff
@@ -55,10 +55,30 @@ pub fn check_success(output: &std::process::ExitStatus) {
...
+    );
+    let output = std::path::Path::new(&location);
+    if !output.parent().unwrap().is_dir() {
+        std::fs::create_dir_all(output.parent().unwrap()).unwrap();
+    }
+    std::fs::write(output, interpretation.to_string()).unwrap();
+}
+
+pub fn produce_interpretation(raw_command_help: &str) {
+    let (cmd, interpretation) = interpret_help_message(raw_command_help);
```
[the remark this answers is not in the document]
Comment (zancas) [reply src/lib.rs:80]: Good points.  I agree. Fixing..

[the remark this answers is not in the document]
Comment (zancas) [reply src/lib.rs]: Fixed.

File: src/lib.rs [hunk 5]
```diff
@@ -95,11 +115,11 @@ fn scrub_result(cmd_name: String, result_data: String) -> String {
     );
         let scrub_3 = scrub_2.replace("(same fields as \"enforce\")", "");
         let scrub_4 = scrub_3.replace(", ...", "");
-        return scrub_4
-    // Note: "xxxx" ID in upgrades. This represents the hash value
-    // of nuparams, for example `5ba81b19`
-    // TODO note: possible need for commas with multiple members of
-    // softforks and upgrades
+        return scrub_4;
```
[the remark this answers is not in the document]
Comment (zancas) [reply src/lib.rs:118]: That's due to RustFmt...   which is confused by the comments that are added at the end of that function definition.

I have a few suggestions for this `fn`, but I think we should wait until we review the PR into zingolabs.

File: src/lib.rs (hunk 1 above)
Comment (zancas) [reply src/lib.rs:773]: That means that `zcash-cli` is already on your path.

File: src/utils/logging.rs [hunk 6]
```diff
@@ -42,6 +30,19 @@ pub fn get_zcashd_version() -> String {
         .to_string()
 }
 
+pub(crate) fn create_version_name() -> String {
+    format!("{}_{}", get_zcashd_version(), QUIZFACE_VERSION)
+}
+
+pub fn create_log_dirs() {
```
[the remark this answers is not in the document]
Comment (zancas) [reply src/utils/logging.rs:37]: At a cost.

File: src/utils/logging.rs [hunk 7]
```diff
@@ -42,6 +30,19 @@ pub fn get_zcashd_version() -> String {
...
+    fs::create_dir_all(Path::new(&name_logdirs().0))
+        .expect("error crating master dir!");
```
[the remark this answers is not in the document]
Comment (zancas) [reply src/utils/logging.rs:39]: Nice catch!

[the remark this answers is not in the document]
Comment (zancas) [reply src/utils/logging.rs]: Fixed

**Comment by dannasessha** [comment]
So all tests pass, but the code still does not compile.

I believe this is because some of the 'gating' code to only try to interpret certain commands was removed.

I suggest we re-apply something like this to see that our code compiles, and thus produces the logs we are finding useful, as well as being functional in a few cases to provide an interface for `typegen`.

**Comment by zancas** [comment]
> I'm not sure if it's possible to see I have started a review before I finish comments, but if not, it would be helpful in the future to have some chatter about force-pushes or continuing development of the branch after it's mentioned in our communication or a review is requested here on github.

OK...  so it seems like we have successfully resolved this issue.   Github preserves relevant knowledge...  comments etc. so we don't need to engage our own protocol.

**Comment by dannasessha** [comment]
Thanks for the convo about all the issues, going to merge this in, with cleanup if needed in a subsequent branch.

**Pull request merged**